refactor(cmd-api-server): pull OAuth2 endpoint scopes from openapi.json

aldousalvarez opened the pull request:
### **Commit** to be reviewed
---
refactor(cmd-api-server): pull OAuth2 endpoint scopes from openapi.json
```
Primary Changes
----------------
1. added OAuth2 security endpoints scopes to openapi.json
2. added a test to make sure if the scopes are indeed getting 
   pulled from the spec file
```
Fixes #2693

**Pull Request Requirements**
- [ ] Rebased onto `upstream/main` branch and squashed into single commit to help maintainers review it more efficient and to avoid spaghetti git commit graphs that obfuscate which commit did exactly what change, when and, why.
- [ ] Have git sign off at the end of commit message to avoid being marked red. You can add `-s` flag when using `git commit` command. You may refer to this [link](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) for more information.
- [ ] Follow the Commit Linting specification. You may refer to this [link](https://www.conventionalcommits.org/en/v1.0.0-beta.4/#specification) for more information. 

**Character Limit**
- [ ] Pull Request Title and Commit Subject must not exceed 72 characters (including spaces and special characters).
- [ ] Commit Message per line must not exceed 80 characters (including spaces and special characters).

**A Must Read for Beginners**
For rebasing and squashing, here's a [must read guide](https://github.com/servo/servo/wiki/Beginner's-guide-to-rebasing-and-squashing) for beginners.

outSH commented on packages/cactus-cmd-api-server/src/test/typescript/unit/get-open-api-spec-v1-oauth2-scopes.test.ts:
```diff
@@ -0,0 +1,165 @@
+}
+
+describe("cmd-api-server:getOpenApiSpecV1Endpoint", () => {
+  const logLevel: LogLevelDesc = "TRACE";
```
Change log level to `info`, dont use `trace` for tests in CI

aldousalvarez replied:
@outSH Got it, just finished updating it.

petermetz commented on packages/cactus-cmd-api-server/src/test/typescript/unit/get-open-api-spec-v1-oauth2-scopes.test.ts:
```diff
@@ -0,0 +1,165 @@
+    expect(res1.data).toBeTruthy();
+
+    console.log("Response data type:", typeof res1.data);
+    console.log("Response data:", res1.data);
+
+    let openApiSpec;
+    try {
+      openApiSpec =
+        typeof res1.data === "string" ? JSON.parse(res1.data) : res1.data;
+    } catch (error) {
+      throw new Error(`Failed to parse OpenAPI spec: ${error.message}`);
+    }
+
+    expect(openApiSpec).toHaveProperty("components");
+    expect(openApiSpec.components).toHaveProperty("securitySchemes");
+
+    const securitySchemes = openApiSpec.components.securitySchemes;
+    expect(securitySchemes).toBeObject();
+
+    const expectedScopes: IExpectedScopes = {
+      "read:health": "Read health information",
+      "read:metrics": "Read metrics information",
+      "read:spec": "Read OpenAPI specification",
+    };
+
+    const securitySchemeNames = Object.keys(securitySchemes);
+
+    securitySchemeNames.forEach((schemeName) => {
+      const scheme = securitySchemes[schemeName];
+      expect(scheme).toHaveProperty("flows");
+      const flows = scheme.flows;
+      expect(flows).toHaveProperty("authorizationCode");
+      const scopes = flows.authorizationCode.scopes as IExpectedScopes;
+
+      Object.keys(expectedScopes).forEach((scope) => {
+        expect(scopes).toHaveProperty(scope);
+        expect(scopes[scope]).toEqual(expectedScopes[scope]);
+      });
+    });
+  });
```
@aldousalvarez This verifies that the OpenAPI .json file has the security scheme declared.  It does not verify that those are being respected by the endpoint implementations themselves. Please update the endpoints touched by this diff to make use of the declared scopes and then write **2 more test cases for each endpoint** in addition this one I'm commenting on. One of them should be the positive test case where with a valid JWT the API client can execute the request and with an invalid JWT the request execution fails with an HTTP 401 unauthorized error.
Note: by test cases I just mean another `it("does something...", async () => {...})` block not a separate file.

petermetz replied:
@aldousalvarez Thank you

aldousalvarez commented:
Already done the requested changes, re-requested for your review. Thank you

petermetz commented on packages/cactus-cmd-api-server/src/test/typescript/unit/get-open-api-spec-v1-oauth2-scopes.test.ts:
```diff
@@ -0,0 +1,209 @@
+    const res1Promise = apiClient.getOpenApiSpecV1();
+    await expect(res1Promise).resolves.toHaveProperty("data.openapi");
+    const res1 = await res1Promise;
+    expect(res1.status).toEqual(200);
+    expect(res1.data).toBeTruthy();
```
@aldousalvarez In this test case it is supposed to fail because you are executing with a JWT that does not have the scope declared (so the JWT has a valid signature but it does not declare the correct role based access control and therefore should not work)

petermetz replied:
@aldousalvarez Thank you for the updates here as well!

petermetz commented on packages/cactus-cmd-api-server/src/test/typescript/unit/get-open-api-spec-v1-oauth2-scopes.test.ts:
```diff
@@ -0,0 +1,239 @@
+    const resPromise = apiClient.getHealthCheckV1({
+      headers: { Authorization: invalidBearerToken },
+    });
+    await expect(resPromise).rejects.toThrow(
```
@aldousalvarez Almost there! Please make sure to explicitly assert that the `status` property of the is 401 or 403 depending on which one is applicable (403 if the token is valid BUT missing the correct scope, 401 otherwise)

petermetz replied:
@aldousalvarez Thank you for the updates!

petermetz commented on packages/cactus-cmd-api-server/src/test/typescript/unit/get-open-api-spec-v1-oauth2-scopes.test.ts:
```diff
@@ -0,0 +1,239 @@
+    });
+    await expect(resPromise).rejects.toThrow(
+      "Request failed with status code 401",
+    );
```
@aldousalvarez Almost there! Please make sure to explicitly assert that the `status` property of the is 401 or 403 depending on which one is applicable (403 if the token is valid BUT missing the correct scope, 401 otherwise)

petermetz replied:
@aldousalvarez Thank you for the updates!

petermetz reviewed (changes requested):
@aldousalvarez Please fix the rest of the failing tests of the cmd-api-server package and also the custom-checks check which also is failing.

aldousalvarez commented:
Hello @petermetz I added another test case where it would be 403, and then I also added that it logs which status code it gets,. Should I also add 403 in the responses on the openapi.json as well? Thanks

petermetz commented:
> Hello @petermetz I added another test case where it would be 403, and then I also added that it logs which status code it gets,. Should I also add 403 in the responses on the openapi.json as well? Thanks

@aldousalvarez Sorry for the slow response! Thank you for the changes and to your question: Yes, please declare the 403 in the openapi.tpl.json spec as well (and make sure to run codegen so that it updates openapi.json also)

aldousalvarez commented:
Hello @petermetz re-requested for review. already done with the requested changes and updated some of the tests with the required scope after running yarn run codegen so the tests would pass. Thank you

petermetz commented on packages/cactus-cmd-api-server/src/test/typescript/unit/get-open-api-spec-v1-oauth2-scopes.test.ts:
```diff
@@ -0,0 +1,274 @@
+      const statusCode = error.response?.status;
+      console.log(`Request failed with status code: ${statusCode}`);
+    }
+  });
```
@aldousalvarez Thank you!

petermetz commented on packages/cactus-cmd-api-server/src/test/typescript/unit/get-open-api-spec-v1-oauth2-scopes.test.ts:
```diff
@@ -0,0 +1,274 @@
+    const resPromise = apiClient.getHealthCheckV1({
+      headers: { Authorization: invalidBearerToken },
+    });
+    await expect(resPromise).rejects.toThrow();
```
@aldousalvarez Thank you for the updates!

petermetz reviewed (approved):
LGTM
